kwheelan/datamade-code-challenge
# Kwheelan solution

## Overview

This PR completes the coding challenge for DataMade's developer position:

- Fixes Docker build error from missing `npm` install statement
- Implements the `parse` method in `parserator_web/views.py`
- Completes the API endpoint, including error handling
- Wires html form to receive address input and send as a parameter to the API
- Displays the results from the API if successful, and an error message if unsuccessful
- Includes 2 unit tests

## Demo

### Success

https://github.com/user-attachments/assets/dfdb1a10-a491-4c8a-afa7-352368bde2cd

### Fail to parse

https://github.com/user-attachments/assets/9719553d-a707-4866-a4b7-751734ece383

## Testing Instructions 

(Excerpted from the ReadMe)

 Development requires a local installation of [Docker](https://docs.docker.com/install/) and [Docker Compose](https://docs.docker.com/compose/install/). These are the only two system-level dependencies you should need.

Once you have Docker and Docker Compose installed, build the application containers:

```
docker-compose build
```

Next, run the app:

```
docker-compose up
``` 

The app will log to the console, and you should be able to visit it at [http://localhost:8000/](http://localhost:8000/).

To test use:

```
docker-compose -f docker-compose.yml -f tests/docker-compose.yml run --rm app
```

As shown above, the test cases are :

 - ✅ Valid: `123 main st chicago il`, which should return `200` from the API and display the fetched data correctly
 - ❌ Invalid: `123 main st chicago il 123 main st`, which should return `422` from the API and display a user-friendly error message.
